### PR TITLE
Added `try_exists()` method to `std::path::Path`

### PR DESCRIPTION
**Pull request by Kixunil**
This method is similar to the existing `exists()` method, except it
doesn't silently ignore the errors, leading to less error-prone code.

This change intentionally does NOT touch the documentation of `exists()`
nor recommend people to use this method while it's unstable.
Such changes are reserved for stabilization to prevent confusing people.

Apart from that it avoids conflicts with #80979.

@joshtriplett requested this PR in [internals discussion](https://internals.rust-lang.org/t/the-api-of-path-exists-encourages-broken-code/13817/25?u=kixunil)

### CODE REVIEW

File: library/std/src/path.rs [hunk 1]
```diff
@@ -2468,6 +2468,32 @@ impl Path {
...
+    /// assert!(Path::new("/root/secret_file.txt").try_exists().is_err());
+    /// ```
+    // TODO: stabilization should modify documentation of `exists()` to recommend this method
+    // instead.
+    #[unstable(feature = "path_try_exists", issue = "none")]
+    #[inline]
+    pub fn try_exists(&self) -> io::Result<bool> {
+        fs::metadata(self)
+            .map(|_| true)
+            .or_else(|error| if error.kind() == io::ErrorKind::NotFound { Ok(false) } else { Err(error) } )
```
Comment (tesuji) [inline library/std/src/path.rs:2494]: Why not use a match expression?
```suggestion
        match fs::metadata(self) {
            Ok(_) => Ok(true),
            Err(e) if e.kind() == io::ErrorKind::NotFound => Ok(false),
            Err(e) => Err(e),
        }
```

Comment (Kixunil) [reply library/std/src/path.rs:2494]: Hmm, good point, I usually use combinators as they tend to be more succinct but match may be better here.

Is there a generally preferred style? Maybe others have a different opinion?

Comment (jplatte) [reply library/std/src/path.rs:2494]: I agree that in this case, a `match` is easier to read.

**Comment by ChrisDenton** [comment]
One thing I'm unclear on is how this should work on other platforms. I'm thinking in particular of Windows which does not have the same folder issue as Linux but has its own way of working. E.g. `GetFileAttributes` can have one of the following outcomes:

* **Success**: the file exists
* **Access denied error**: the file exists but the attributes can't be read.
* **File not found error**: the file does not exist
* **Other error**: Something very unexpected happened.

If a file does exist but reading its metadata fails, then should it return an error? Or `true`?

**Comment by Kixunil** [comment]
@ChrisDenton access denied is just one (obvious) example, there can be various other reasons depending on the underlying OS.

> If a file does exist but reading its metadata fails

If `metadata()` fails with any error other than `NotFound` then *we don't know* if it exists. The only reasonable return value is thus `Err`. That was the entire point of previous discussion.

**Comment by ChrisDenton** [comment]
> there can be various other reasons depending on the underlying OS

Yes the behaviour can be very OS specific, which is the point I was trying to make using Windows as an example. But if the idea is to return `Ok(true)` only if the files' metadata is readable then I do think the documentation should make that clear in some way.

File: library/std/src/path.rs [hunk 2]
```diff
@@ -2468,6 +2468,34 @@ impl Path {
...
     }
 
+    /// Returns `Ok(true)` if the path points at an existing entity.
+    ///
+    /// This function will traverse symbolic links to query information about the
+    /// destination file. In case of broken symbolic links this will return `Ok(false)`.
+    ///
+    /// As opposed to the `exists()` method, this one doesn't silently ignore errors
+    /// unrelated to the path not existing. (E.g. it will return `Err(_)` in case of permission
+    /// denied on some of the paret directories.)
```
Comment (kennytm) [inline library/std/src/path.rs:2478]:
```suggestion
    /// denied on some of the parent directories.)
```

**Comment by joshtriplett** [comment]
This looks good to me (with the suggested fix applied).

**Comment by Kixunil** [comment]
@joshtriplett I guess you meant fix that `match`, so I did. Happy to have helped!

Should I open the tracking issue already or wait for something first?

File: library/std/src/path.rs (hunk 1 above)
Comment (Kixunil) [reply library/std/src/path.rs]: Applied.

**Comment by Kixunil** [comment]
@ChrisDenton do I understand correctly that Windows has some other function to check if file exists and also a function to read metadata and the latter could fail even if the former doesn't?

**Comment by ChrisDenton** [comment]
> @ChrisDenton do I understand correctly that Windows has some other function to check if file exists and also a function to read metadata and the latter could fail even if the former doesn't?

There are two issues I think. The first one is really about how `exists` is implemented in terms of `fs::metadata`. The Windows implementation of `fs::metadata` involves two or more OS calls (first to open the file and then to query metadata). Querying metadata [can fail](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfileinformationbyhandle#remarks) but we only need to open the file to confirm the file exists.

The second issue is that `exists` and `try_exists` can fail for files that do exist and that the error itself indicates the file does indeed exist. For example, when accessing a local file both `ERROR_SHARING_VIOLATION` and `ERROR_ACCESS_DENIED` indicate the file exists but cannot be opened. Note that, unlike Linux, opening won't be denied by permissions of the parent directory because only the file's permissions are taking into account.

I don't think we necessarily need to handle or document all possibilities but I think there could be some more robust wording about what an error can mean. Though I'm struggling with how to phrase it.

**Comment by Kixunil** [comment]
@ChrisDenton thank you for explaining!

I'd actually prefer to handle those cases properly. If the function name indicates checking of existence of a file it should be as close to reality as possible. It also avoids wordy documentation of such behavior.

File: library/std/src/path.rs [hunk 3]
```diff
@@ -2468,6 +2468,36 @@ impl Path {
...
+    /// ```no_run
+    /// #![feature(path_try_exists)]
+    ///
+    /// use std::path::Path;
+    /// assert!(!Path::new("does_not_exist.txt").try_exists().expect("Can't check existence of file does_not_exist.txt"));
+    /// assert!(Path::new("/root/secret_file.txt").try_exists().is_err());
+    /// ```
+    // FIXME: stabilization should modify documentation of `exists()` to recommend this method
+    // instead.
+    #[unstable(feature = "path_try_exists", issue = "none")]
```
Comment (kennytm) [inline library/std/src/path.rs:2491]: @rust-lang/libs if no one opposes i'll r+ after the tracking issue is filed.

Comment (BurntSushi) [reply library/std/src/path.rs:2491]: LGTM.

Comment (Kixunil) [reply library/std/src/path.rs]: Took it as "yes" answer to my question about opening the tracking issue, so I opened it (#83186) and filled.

**Pull request merged**